Allow SSIZE_T in addition to ssize_t

c72578 opened the pull request:
The Windows Data Type `SSIZE_T` is declared in `BaseTsd.h`
However, it is written in capital letters

- Fixes e.g. the following false positive:
```console
  (portability) %zd in format string (no. 1) requires 'ssize_t' but the
  argument type is 'SSIZE_T {aka signed long long}'.
  [invalidPrintfArgType_sint]
```

orbitcowboy commented:
Thank you for this fix. Types should not be hard coded. Instead we are using library files. Please take a look at cfg/windows.cfg. 
In addition a test case would be nice. The corresponding test files are stored at test/cfg/windows.cpp.

c72578 commented:
Thanks for your feedback
Concerning cfg/windows.cfg, what do you think about the following:
```diff
diff --git a/cfg/windows.cfg b/cfg/windows.cfg
index 2ca20db64..53be0f5d8 100644
--- a/cfg/windows.cfg
+++ b/cfg/windows.cfg
@@ -832,11 +832,11 @@
     <platform type="win32W"/>
     <platform type="win64"/>
   </platformtype>
-  <platformtype name="SSIZE_T" value="long">
+  <platformtype name="SSIZE_T" value="ssize_t">
     <platform type="win32A"/>
     <platform type="win32W"/>
   </platformtype>
-  <platformtype name="SSIZE_T" value="long">
+  <platformtype name="SSIZE_T" value="ssize_t">
     <long/>
     <platform type="win64"/>
   </platformtype>
```

orbitcowboy commented:
> Thanks for your feedback
> Concerning cfg/windows.cfg, what do you think about the following:
> 
> ```diff
> diff --git a/cfg/windows.cfg b/cfg/windows.cfg
> index 2ca20db64..53be0f5d8 100644
> --- a/cfg/windows.cfg
> +++ b/cfg/windows.cfg
> @@ -832,11 +832,11 @@
>      <platform type="win32W"/>
>      <platform type="win64"/>
>    </platformtype>
> -  <platformtype name="SSIZE_T" value="long">
> +  <platformtype name="SSIZE_T" value="ssize_t">
>      <platform type="win32A"/>
>      <platform type="win32W"/>
>    </platformtype>
> -  <platformtype name="SSIZE_T" value="long">
> +  <platformtype name="SSIZE_T" value="ssize_t">
>      <long/>
>      <platform type="win64"/>
>    </platformtype>
> ```
According to https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types SSIZE_T  is defined as typedef LONG_PTR SSIZE_T;

c72578 commented:
Yes, that's the correct link to the Windows Data Types.
However, currently "ssize_t" is hardcoded in  **`lib/checkio.cpp`** using small letters:
L844:
`if (!typesMatch(argInfo.typeToken->originalName(), "ssize_t"))`
L1098:
`if (!typesMatch(argInfo.typeToken->originalName(), "ssize_t"))`

So, at the moment it will only match, if `ssize_t` is in small letters.
The PR is intended to address this.

danmar commented:
To me it sounds ok to use standard type ssize_t to define it.. that seems to match.

danmar commented:
however the intention of the checker is to be very strict. It probably wants to warn if you pass a SSIZE_T value using a ssize_t modifier even if they match.

danmar commented:
@IOBYTE what do you think about `SSIZE_T` / `ssize_t` mismatch?

IOBYTE commented:
There is no good solution because SSIZE_T is not ssize_t. You can't use` <platformtype name="SSIZE_T" value="ssize_t">` because a platform type of a platform type doesn't work.

You can use `<define name="SSIZE_T" value="ssize_t"/>` but it's not technically correct because SSIZE_T is not defined as ssize_t and the error message will not show SSIZE_T.  It will show ssize_t.

```
~/cppcheck$ cat ssize_t.cpp
int main()
{
    SSIZE_T s = 0;
    printf("%zu", s);
}
~/cppcheck$ ./cppcheck --debug ssize_t.cpp --enable=all --inconclusive --library=windows --platform=win32A
Checking ssize_t.cpp ...


##file ssize_t.cpp
1: int main ( )
2: {
3: $long s@1 ; s@1 = 0 ;
4: printf ( "%zu" , s@1 ) ;
5: }



##Value flow
Line 3
  0 always 0
Line 4
  "%zu" always "%zu"
  s always 0
ssize_t.cpp:4:5: portability: %zu in format string (no. 1) requires 'size_t' but the argument type is 'ssize_t {aka signed long}'. [invalidPrintfArgType_uint]
    printf("%zu", s);
    ^

```

c72578 commented:
@IOBYTE, what is your opinion concerning this PR in its initial form: Allowing both in **`lib/checkio.cpp`**: ssize_t and SSIZE_T.

Here is an example test file **`ssize_t.c`**
```c
#include <stdio.h>
#ifdef _MSC_VER
#include <BaseTsd.h>
typedef SSIZE_T ssize_t;
#endif

int main()
{
    /* ssize_t is signed */
    ssize_t s = -2;
    /* Correct format specifier %zd */
    printf("Using %%zd: s=%zd\n", s);
    /* Incorrect format specifier %zu */
    printf("Using %%zu: s=%zu\n", s);
}
```
1.) The current output of cppcheck master is the following:
```console
[ssize_t.c:12] (portability) %zd in format string (no. 1) requires 'ssize_t' but the argument type is 'SSIZE_T {aka signed long long}'. [invalidPrintfArgType_sint]
[ssize_t.c:14] (portability) %zu in format string (no. 1) requires 'size_t' but the argument type is 'SSIZE_T {aka signed long long}'. [invalidPrintfArgType_uint]
```
2.) After the fix, when this PR is used:
```console
[ssize_t.c:14] (portability) %zu in format string (no. 1) requires 'size_t' but the argument type is 'SSIZE_T {aka signed long long}'. [invalidPrintfArgType_uint]
```
So, the output of 2.) is as expected

IOBYTE commented:
It's the easiest way to fix this false positive. We already have a precedent for hard coding windows types to prevent false positives like this in this check but that doesn't make it right.  You should add an isWindows check because this is a windows only hack. You should also consider adding SIZE_T support and especially add a lot of false positive and false negative tests to verify that it works as expected on windows and non-windows platforms.

danmar commented:
yes I also think so. we'll accept hardcoding in this case.

danmar commented:
> You should add an isWindows check because this is a windows only hack. You should also consider adding SIZE_T support and especially add a lot of false positive and false negative tests to verify that it works as expected on windows and non-windows platforms.

For information.. I want to have a isWindows check. And I want to see some tests to avoid regressions.

c72578 commented:
OK, so I would add some tests in **`test/testio.cpp`** to `testMicrosoftPrintfArgument()` and `testMicrosoftScanfArgument()`

c72578 commented:
The PR has been updated:
The `isWindows` checks have been added, as well as some tests.

IOBYTE commented on test/testio.cpp:
```diff
@@ -4142,6 +4142,20 @@ class TestIO : public TestFixture {
                       "[test.cpp:2]: (warning) %d in format string (no. 2) requires 'int' but the argument type is 'const void *'.\n", errout.str());
 
+        check("void foo() {\n"
+              "typedef SSIZE_T ssize_t;\n" // In MSVC, SSIZE_T is available in capital letters using #include <BaseTsd.h>
```
Good start.

You don't need the typedef. It's already defined in windows.cfg.  I'm not sure which one takes precedent. You should have tests with and without the typedef just to be safe.  There is also the problem of someone adding: `typedef SSIZE_T char;`.  Do we want to validate SSIZE_T is really ssize_t?

The problem is windows.cfg is always included and you want to test with and without it.  One way to fix that is to create two different settings (windows and not windows) in TestIO::run and select which one in TestIO::check based on platform type.  Unfortunately that doesn't let you test a windows platform without windows.cfg. This is going to be tricky to get right. This is a preexisting problem that should be worked out first.

You only added false positive tests.  You need %zu and others for false negatives.

You need to test other platforms besides windows.

You have a big matrix of configurations (with and without typedef, with and without windows.cfg, windows and non windows platforms, multiple format specifiers).

Sorry but doing this right is going to be complicated.

c72578 commented:
@IOBYTE thanks for the feedback.
I have updated the PR in the meantime, addressing the points, which are more straight forward:
- `typedef` is not used in general any more in the tests. A separate test using `typedef SSIZE_T ssize_t;` has been added
- `%zu` as a false negative test has bee added
- Tests using multiple format specifiers have been added
- A non Windows platform has been added (Unix64).

danmar commented:
Thanks!
